File: lachancuoicgoi_flutter/lib/analysis/l2/g_detection/g_pattern_matcher.dart
import 'dart:math' as math;

import '../../analysis_result.dart';
import 'g_flash.dart';
import 'g_models.dart';

class GPatternMatcher {
  GPatternMatcher._();

  static final Map<String, String> _normalizedKeywordCache = <String, String>{};

  static List<MatchedPattern> matchPatterns(
    List<String> tokens,
    List<ScamPattern> patterns,
    Set<KeywordMatch> keywordMatches,
  ) {
    final results = <MatchedPattern>[];
    final matchesByIndex = <int, List<KeywordMatch>>{};
    for (final match in keywordMatches) {
      matchesByIndex.putIfAbsent(match.startIndex, () => []).add(match);
    }

    for (final pattern in patterns) {
      final capturedElements = <String>[];
      if (_matchSinglePattern(
        tokens,
        pattern,
        matchesByIndex,
        capturedElements,
      )) {
        results.add(
          MatchedPattern(
            patternId: pattern.id,
            matchedElements: List<String>.of(capturedElements),
            score: pattern.riskBonus,
          ),
        );
      }
    }

    return results;
  }

  static bool _matchSinglePattern(
    List<String> tokens,
    ScamPattern pattern,
    Map<int, List<KeywordMatch>> matchesByIndex,
    List<String> capturedElements,
  ) {
    final firstElement = pattern.template.isEmpty
        ? null
        : pattern.template.first;
    if (firstElement == null) return false;

    for (var i = 0; i < tokens.length; i++) {
      if (_checkElementAt(i, firstElement, tokens, matchesByIndex)) {
        capturedElements
          ..clear()
          ..add(tokens[i]);
        if (_matchRemainingSequence(
          i,
          1,
          tokens,
          pattern,
          matchesByIndex,
          capturedElements,
        )) {
          return true;
        }
      }
    }

    capturedElements.clear();
    return false;
  }

  static bool _matchRemainingSequence(
    int currentIndex,
    int patternIndex,
    List<String> tokens,
    ScamPattern pattern,
    Map<int, List<KeywordMatch>> matchesByIndex,
    List<String> capturedElements,
  ) {
    if (patternIndex >= pattern.template.length) return true;

    final targetElement = pattern.template[patternIndex];
    final searchStart = currentIndex + 1;
    final searchEnd = math.min(tokens.length, searchStart + pattern.maxGap + 1);

    for (var nextIndex = searchStart; nextIndex < searchEnd; nextIndex++) {
      if (_checkElementAt(nextIndex, targetElement, tokens, matchesByIndex)) {
        capturedElements.add(tokens[nextIndex]);
        if (_matchRemainingSequence(
          nextIndex,
          patternIndex + 1,
          tokens,
          pattern,
          matchesByIndex,
          capturedElements,
        )) {
          return true;
        }
        capturedElements.removeLast();
      }
    }

    return false;
  }

  /// Check if element matches at given index.
  /// Returns true if element matches, false otherwise.
  /// For PatternKeyword, supports multi-token keywords (e.g. "chuyen khoan").
  static bool _checkElementAt(
    int index,
    PatternElement element,
    List<String> tokens,
    Map<int, List<KeywordMatch>> matchesByIndex,
  ) {
    if (index >= tokens.length) return false;
    final result = _tryConsumeElement(index, element, tokens, matchesByIndex);
    return result != -1;
  }

  static int _tryConsumeElement(
    int index,
    PatternElement element,
    List<String> tokens,
    Map<int, List<KeywordMatch>> matchesByIndex,
  ) {
    if (index >= tokens.length) return -1;

    return switch (element) {
      PatternKeyword(:final value) => _matchKeyword(tokens, index, value),
      PatternCategory(:final categoryName) =>
        matchesByIndex[index]?.any(
              (match) =>
                  match.category.toLowerCase() == categoryName.toLowerCase(),
            ) ==
            true
            ? index
            : -1,
      PatternWildcard() => index,
    };
  }

  /// Match keyword tokens starting at [startIndex] in the input tokens.
  /// Returns the index of the last matched token, or -1 if no match.
  /// For multi-token keywords (e.g. "chuyen khoan"), checks sequential tokens.
  /// ✓ BUG #2 FIX: Now correctly checks ALL keyword tokens, not just the first.
  static int _matchKeyword(List<String> tokens, int startIndex, String rawKeyword) {
    final normalized = _normalizedKeywordCache.putIfAbsent(
      rawKeyword,
      () => GFlash.tokenize(rawKeyword).join(' '),
    );
    final keywordTokens = normalized
        .split(' ')
        .where((part) => part.isNotEmpty)
        .toList();
    if (keywordTokens.isEmpty) return -1;

    // Check all keyword tokens match sequentially from startIndex
    if (startIndex + keywordTokens.length > tokens.length) return -1;
    for (var i = 0; i < keywordTokens.length; i++) {
      if (tokens[startIndex + i] != keywordTokens[i]) return -1;
    }
    // Return index of last matched token
    return startIndex + keywordTokens.length - 1;
  }
}

File: lachancuocgoi_kotline/app/src/main/java/com/example/lachancuocgoi/Analysis/L2/GDetection/GPatternMatcher.kt
package com.example.lachancuocgoi.Analysis.L2.GDetection

import com.example.lachancuocgoi.Analysis.KeywordMatch
import android.util.Log
import kotlin.math.min

/**
 * [B3 UPGRADE] Engine tìm kiếm mẫu câu trong văn bản đã token hóa.
 * [FIX] Pattern keyword được normalize qua GFlash trước khi so khớp với tokens.
 */
object GPatternMatcher {
    private const val TAG = "GPatternMatcher"

    /**
     * Cache normalized keywords để tránh re-tokenize mỗi lần call.
     * Key = raw keyword string, Value = first token (hoặc joined tokens).
     */
    private val normalizedKeywordCache = mutableMapOf<String, String>()

    private fun normalizeKeyword(raw: String): String {
        return normalizedKeywordCache.getOrPut(raw) {
            GFlash.tokenize(raw).joinToString(" ")
        }
    }

    /**
     * Tìm kiếm các mẫu câu trong văn bản.
     * keywordMatches: Set các keyword L2 đã tìm thấy trong transcript (với index)
     */
    fun matchPatterns(
        tokens: List<String>,
        patterns: List<ScamPattern>,
        keywordMatches: Set<KeywordMatch>
    ): List<MatchedPattern> {
        val results = mutableListOf<MatchedPattern>()

        // Build map for efficient lookup: Index -> List<KeywordMatch> starting at this index
        val matchesByIndex = keywordMatches.groupBy { it.startIndex }

        patterns.forEach { pattern ->
            val capturedElements = mutableListOf<String>()
            if (matchSinglePattern(tokens, pattern, matchesByIndex, capturedElements)) {
                results.add(
                    MatchedPattern(
                        patternId = pattern.id,
                        matchedElements = capturedElements.toList(),
                        score = pattern.riskBonus
                    )
                )
                Log.d(TAG, "Pattern matched: ${pattern.id} → [${capturedElements.joinToString(" ... ")}]")
            }
        }

        return results
    }

    private fun matchSinglePattern(
        tokens: List<String>,
        pattern: ScamPattern,
        matchesByIndex: Map<Int, List<KeywordMatch>>,
        capturedElements: MutableList<String>
    ): Boolean {
        val firstElement = pattern.template.firstOrNull() ?: return false

        for (i in tokens.indices) {
            if (checkElementAt(i, firstElement, tokens, matchesByIndex)) {
                capturedElements.clear()
                capturedElements.add(tokens[i])
                if (matchRemainingSequence(i, 1, tokens, pattern, matchesByIndex, capturedElements)) {
                    return true
                }
            }
        }
        capturedElements.clear()
        return false
    }

    private fun matchRemainingSequence(
        currentIndex: Int,
        patternIndex: Int,
        tokens: List<String>,
        pattern: ScamPattern,
        matchesByIndex: Map<Int, List<KeywordMatch>>,
        capturedElements: MutableList<String>
    ): Boolean {
        if (patternIndex >= pattern.template.size) return true

        val targetElement = pattern.template[patternIndex]
        val searchStart = currentIndex + 1
        val searchEnd = min(tokens.size, searchStart + pattern.maxGap + 1)

        for (nextIndex in searchStart until searchEnd) {
            if (checkElementAt(nextIndex, targetElement, tokens, matchesByIndex)) {
                capturedElements.add(tokens[nextIndex])
                if (matchRemainingSequence(nextIndex, patternIndex + 1, tokens, pattern, matchesByIndex, capturedElements)) {
                    return true
                }
                capturedElements.removeAt(capturedElements.lastIndex) // Backtrack
            }
        }

        return false
    }

    private fun checkElementAt(
        index: Int,
        element: PatternElement,
        tokens: List<String>,
        matchesByIndex: Map<Int, List<KeywordMatch>>
    ): Boolean {
        if (index >= tokens.size) return false

        return when (element) {
            is PatternElement.Keyword -> {
                // [FIX - Critical] Normalize the pattern keyword via GFlash before comparing.
                // tokens[] are already normalized (lowercase, no diacritics, slang-replaced).
                // Previously element.value was raw Vietnamese → never matched normalized tokens.
                // Now we normalize the keyword to match, and also try single-token prefix match
                // for multi-word keywords stored as a single phrase.
                val normalizedSingle = normalizeKeyword(element.value)
                // normalizedSingle may be multi-word (e.g. "chuyen tien"), but tokens[index]
                // is a single token. So we only match if normalizedSingle is exactly one token.
                val normalizedTokens = normalizedSingle.split(" ").filter { it.isNotBlank() }
                if (normalizedTokens.size == 1) {
                    tokens[index] == normalizedTokens[0]
                } else {
                    // Multi-word keyword: match the first token here; remaining tokens will be
                    // matched via sequential wildcard steps added by GDetectionEngine at load time.
                    // For now fall back to prefix-match of the first token.
                    tokens[index] == (normalizedTokens.firstOrNull() ?: "")
                }
            }
            is PatternElement.Category -> {
                val matchesHere = matchesByIndex[index]
                matchesHere?.any { it.category.equals(element.categoryName, ignoreCase = true) } == true
            }
            is PatternElement.Wildcard -> true
        }
    }
}
